### PR TITLE
Loosen up typed container casting checks

### LINKED ISSUE #4307: Can't access Typed Dict with (int32, int32) keys in the python interpreter
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

<details>
<summary> output of numba -s </summary>

```bash
$ numba -s
System info:
--------------------------------------------------------------------------------
__Time Stamp__
2019-07-11 09:37:23.968590

__Hardware Information__
Machine                                       : x86_64
CPU Name                                      : skylake
Number of accessible CPU cores                : 8
Listed accessible CPUs cores                  : 0-7
CFS restrictions                              : None
CPU Features                                  : 
64bit adx aes avx avx2 bmi bmi2 clflushopt cmov cx16 f16c fma fsgsbase invpcid
lzcnt mmx movbe pclmul popcnt prfchw rdrnd rdseed rtm sahf sgx sse sse2 sse3
sse4.1 sse4.2 ssse3 xsave xsavec xsaveopt xsaves

__OS Information__
Platform                                      : Linux-4.18.0-25-generic-x86_64-with-Ubuntu-18.04-bionic
Release                                       : 4.18.0-25-generic
System Name                                   : Linux
Version                                       : #26~18.04.1-Ubuntu SMP Thu Jun 27 07:28:31 UTC 2019
OS specific info                              : Ubuntu18.04bionic
glibc info                                    : glibc 2.15

__Python Information__
Python Compiler                               : GCC 5.4.0 20160609
Python Implementation                         : CPython
Python Version                                : 2.7.16
Python Locale                                 : en_GB UTF-8

__LLVM information__
LLVM version                                  : 8.0.0

__CUDA Information__
Error: CUDA device intialisation problem. Error at driver init: 
[100] Call to cuInit results in CUDA_ERROR_NO_DEVICE:
Error class: <class 'numba.cuda.cudadrv.error.CudaSupportError'>

__ROC Information__
ROC available                                 : False
Error initialising ROC due to                 : No ROC toolchains found.
No HSA Agents found, encountered exception when searching:
Error at driver init: 
NUMBA_HSA_DRIVER /opt/rocm/lib/libhsa-runtime64.so is not a valid file path.  Note it must be a filepath of the .so/.dll/.dylib or the driver:

__SVML Information__
SVML state, config.USING_SVML                 : False
SVML library found and loaded                 : False
llvmlite using SVML patched LLVM              : True
SVML operational                              : False

__Threading Layer Information__
TBB Threading layer available                 : True
OpenMP Threading layer available              : False
+--> Disabled due to                          : Unknown import problem.
Workqueue Threading layer available           : True

__Numba Environment Variable Information__
None set.

__Conda Information__
Conda not present/not working.
Error was [Errno 2] No such file or directory

--------------------------------------------------------------------------------
If requested, please copy and paste the information between
the dashed (----) lines, or from a given specific section as
appropriate.

=============================================================
IMPORTANT: Please ensure that you are happy with sharing the
contents of the information present, any information that you
wish to keep private you should remove before sharing.
=============================================================
```

</details>

I created a jitted numba function that created a typed dict keyed on an (int32, int32) tuple. numba successfully creates and returns this dict, but there seems to a problem when accessing the dict within the python interpreter. However, a (int64, int64) tuple works successfully

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from pprint import pprint

import numba
import numpy as np


@numba.njit(nogil=True, cache=True)
def fn(ant1, ant2, data):
    data_avg = {}

    for r in range(ant1.shape[0]):
        a1 = ant1[r]
        a2 = ant2[r]

        key = (a1, a2)

        if key in data_avg:
            avg = data_avg[key]
            avg += data[r]
        else:
            avg = data.dtype.type(data[r])
            data_avg[key] = avg

    return data_avg


# 64 bit indices in ant1 + ant2, this is fine
ant1, ant2 = np.triu_indices(7, 1)
data = np.random.random(ant1.shape)
res = fn(ant1, ant2, data)

for k, v in res.items():
    print(k, v)

# 32 bit indices in ant1 + ant2, this fails!
ant1, ant2 = (a.astype(np.int32) for a in np.triu_indices(7, 1))
data = np.random.random(ant1.shape)
res = fn(ant1, ant2, data)

for k, v in res.items():
    print(k, v)
```

It produces the following sort of traceback:

```traceback
$ python test_numba_int64.py 
(0, 1) 0.53562473169
(0, 2) 0.323425964623
(0, 3) 0.602906503761
(0, 4) 0.0706116917503
(0, 5) 0.698970661917
(0, 6) 0.598562358806
(1, 2) 0.488224070915
(1, 3) 0.641514625944
(1, 4) 0.356949910527
(1, 5) 0.68163141825
(1, 6) 0.702238307709
(2, 3) 0.524306389349
(2, 4) 0.259019818589
(2, 5) 0.454954563449
(2, 6) 0.887542275029
(3, 4) 0.252880237227
(3, 5) 0.43829949257
(3, 6) 0.985091314436
(4, 5) 0.433945981699
(4, 6) 0.185182054301
(5, 6) 0.25504427879
Traceback (most recent call last):
  File "test_numba_int64.py", line 42, in <module>
    for k, v in res.items():
  File "/home/sperkins/venv/xms-plots/lib/python2.7/_abcoll.py", line 414, in items
    return [(key, self[key]) for key in self]
  File "/home/sperkins/venv/xms-plots/local/lib/python2.7/site-packages/numba/typed/typeddict.py", line 137, in __getitem__
    return _getitem(self, key)
  File "/home/sperkins/venv/xms-plots/local/lib/python2.7/site-packages/numba/dispatcher.py", line 351, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/sperkins/venv/xms-plots/local/lib/python2.7/site-packages/numba/dispatcher.py", line 318, in error_rewrite
    reraise(type(e), e, None)
  File "<string>", line 2, in reraise
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<intrinsic _cast>) with argument(s) of type(s): (tuple(int64 x 2), typeref[tuple(int32 x 2)])
 * parameterized
In definition 0:
    TypingError: cannot safely cast tuple(int64 x 2) to tuple(int32 x 2)
In definition 1:
    TypingError: cannot safely cast tuple(int64 x 2) to tuple(int32 x 2)
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<intrinsic _cast>)
[2] During: typing of call at /home/sperkins/venv/xms-plots/local/lib/python2.7/site-packages/numba/dictobject.py (873)


File "../../../venv/xms-plots/local/lib/python2.7/site-packages/numba/dictobject.py", line 873:
    def impl(d, key):
        castedkey = _cast(key, keyty)
        ^

[1] During: typing of intrinsic-call at /home/sperkins/venv/xms-plots/local/lib/python2.7/site-packages/numba/typed/typeddict.py (35)

File "../../../venv/xms-plots/local/lib/python2.7/site-packages/numba/typed/typeddict.py", line 35:
def _getitem(d, key):
    return d[key]
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/dev/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new

```

### PR DESCRIPTION
**Pull request by sklam**
Fix #4307.

- Keep raising exceptions for cross domain cast for numbers.
- Warn but permit on other type of unsafe cast.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. The current code is consistently failing on Python 3 with:
```
FAIL: test_cast_fail_int_unicode (numba.tests.test_listobject.TestItemCasting)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_listobject.py", line 1455, in test_cast_fail_int_unicode
    str(raises.exception),
AssertionError: 'cannot safely cast int32 to unicode_type' not found in 'Failed in nopython mode pipeline (step: nopython frontend)\nInternal error at <numba.typeinfer.CallConstraint object at 0x7f05a8648cf8>.\nFailed in nopython mode pipeline (step: nopython mode backend)\nCannot cast int32 to unicode_type: %".21" = load i32, i32* %"item"\n\nFile "numba/listobject.py", line 432:\n    def impl(l, item):\n        casteditem = _cast(item, itemty)\n        ^\n\n[1] During: lowering "$0.4 = call $0.1(item, $0.3, args=[Var(item, /home/vsts/work/1/s/numba/listobject.py (432)), Var($0.3, /home/vsts/work/1/s/numba/listobject.py (432))], func=$0.1, kws=[], vararg=None)" at /home/vsts/work/1/s/numba/listobject.py (432)\n[2] During: resolving callee type: BoundFunction((<class \'numba.types.containers.ListType\'>, \'append\') for ListType[unicode_type])\n[3] During: typing of call at /home/vsts/work/1/s/numba/tests/test_listobject.py (1449)\n\nEnable logging at debug level for details.\n\nFile "numba/tests/test_listobject.py", line 1449:\n        def foo():\n            <source elided>\n            l = listobject.new_list(types.unicode_type)\n            l.append(int32(0))\n            ^\n'
```

File: numba/typedobjectutils.py [hunk 1]
```diff
@@ -39,18 +40,34 @@ def _sentry_safe_cast(fromty, toty):
     """Check and raise TypingError if *fromty* cannot be safely cast to *toty*
     """
     tyctxt = cpu_target.typing_context
+    fromty, toty = map(types.unliteral, (fromty, toty))
     by = tyctxt.can_convert(fromty, toty)
+
+    def warn():
+        m = 'unsafe cast from {} to {}. Precision maybe lost.'
```
Comment (stuartarchibald) [inline numba/typedobjectutils.py]:
```suggestion
        m = 'unsafe cast from {} to {}. Precision may be lost.'
```

Comment (stuartarchibald) [reply numba/typedobjectutils.py]: I also wonder if this should not refer to precision as down casting `int`'s may end up with overflow etc.

File: numba/typedobjectutils.py [hunk 2]
```diff
@@ -39,18 +40,34 @@ def _sentry_safe_cast(fromty, toty):
...
-        if isinstance(fromty, types.Float) and isinstance(toty, types.Float):
+            warn()
+        elif isflt(fromty) and isflt(toty):
             # Accept if floats to floats
-            return
-        raise TypingError('cannot safely cast {} to {}'.format(fromty, toty))
+            warn()
+        elif not isinstance(toty, types.Number):
+            # Non-numbers
+            warn()
```
Comment (stuartarchibald) [inline numba/typedobjectutils.py]: I think `(complex64, complex128)` casting should work ok but warn and not be a hard error?

Comment (sklam) [reply numba/typedobjectutils.py]: good catch. fixe in https://github.com/numba/numba/pull/4330/commits/91aef6bfff8d03d0d22d65745e7732a15611d056

File: numba/tests/test_typedobjectutils.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            (types.float32, types.float64),
+            # complex cases
+            (types.complex64, types.complex128),
+            # int to float cases
+            (types.int32, types.float64),
+            (types.uint8, types.float32),
+            # float to complex cases
+            (types.float32, types.complex128),
+            (types.float64, types.complex128),
+            # tuple-of-ints to tuple-of-floats,
```
Comment (stuartarchibald) [inline numba/tests/test_typedobjectutils.py]:
```suggestion
            # tuple-of-ints to tuple-of-ints,
```

**Comment by stuartarchibald** [comment]
Thanks for the fixes, there's a comment with a copy-paste error (noted above), else this is good to go.

**Pull request merged**